### DISCUSSION TITLE
Building resiliency to overload and request throttling by istio sidecar when querying cache and expensive service are involved

### DISCUSSION THREAD

**Discussion by andrii-korotkov-verkada** [istio/istio#49470] [Q&A] [answered]
Imagine a deployment with a main container and istio sidecar. The main container has a logic to query cache (e.g. Redis) and query another service in case there's no cached entry or cache request failed. Querying that another service is expensive. Normally the cache calls succeed often enough so that the load on another service is reasonable.

Now suppose istio gets overloaded and start throttling some requests. In particular, a part of requests to the cache would be affected. This, in turn, can lead to more requests to another service, potentially overloading it as well. How to build resiliency for such scenarios? Ideally the requests to the cache should be prioritized to not be throttled, so that the load on another service is reduced and there's less chance of a cascading failure. But I haven't figured out a way to do that except setting like 10 retries on requests to the cache and/or having rate limit for communication between services.

I'd be grateful for any suggestions on how to deal with the described situation and make the ecosystem more resilient. Thank you.

**Comment by hanxiaop** [comment, marked as answer]
https://istio.io/latest/docs/tasks/traffic-management/circuit-breaking/ this can be a helpful task for you.

**Reply by andrii-korotkov-verkada** [reply]
Thanks! For istio-based circuit-breaking to work, do both the client and destination services need to have sidecar, or it's enough for a destination to have it? I'm asking since multiple services can benefit from having a circuit breaker and I wonder if having a mesh is a requirement or a partial mesh would also work.

**Reply by hanxiaop** [reply]
The circuit-breaking is configured in the CDS, and if your destination service includes the sidecar, it will enforce the policy. That is to say, if you only use circuit-breaking, a partial mesh should work.
